server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -694,6 +708,17 @@ public Version getCreationVersion() {
         return indexCreatedVersion;
     }
 
+    /**
+     * Return the {@link Version} that this index provides compatibility for.
+     * This is typically compared to the {@link Version#minimumIndexCompatibilityVersion()} to figure out whether the index can be handled
```

Comment:
This is likely mostly my uncertainty on the project and intentions. It nags me a little that we instantiate the new mappings/metadata on a specific version and then guarantee to be able to use that on the next major, but no longer than that. Both aspects of that seems wrong and should really be tied to the created-version (if version 10 can read 5 indices, there is no need to do anything about an index restored in 7.x during upgrade to 8, 9 and 10).

I suppose we could upgrade metadata/mappings for every major. And also on minors in case we change how we derive the metadata/mapping. Is that part of the intention? It makes me wonder if we could do this completely dynamically (even at the shard level to support mixed clusters), but that also sounds complex.

Reply:
> I suppose we could upgrade metadata/mappings for every major. And also on minors in case we change how we derive the metadata/mapping. Is that part of the intention?

Correct, that's the intention with the compatibility version. Similarly to the created-version, it defines how "old" the index metadata is and what legacy stuff from the past it might contain. At the same time, I envision this to be at least auto-upgraded every major (preserving / remapping as much of the original metadata as possible). This is why again we have the natural N-1 compatibility with `index.version.compatibility`.

Note that as we preserve the original mapping as well (under _meta/legacy-mappings), it might make sense to remap across minors even to make use of some new features (e.g. doc-value-only fields for new field types).